### PR TITLE
Base Token Bridge

### PR DESCRIPTION
**Pull request by telome**

### CODE REVIEW

File: src/L1TokenBridge.sol
Comment (oldchili) [inline src/L1TokenBridge.sol]: I think this might work even if `_remoteToken` is address(0), so might work even if a `_localToken` is not registered.

File: src/L1TokenBridge.sol
Comment (oldchili) [inline src/L1TokenBridge.sol:190]: Why does the sender have to be an EOA?
How would it work with relaying vested tokens to an L2 farm?

Comment (oldchili) [reply src/L1TokenBridge.sol:190]: Ah nvm, for some reason I thought the other function is internal.

File: src/L1TokenBridge.sol
Comment (telome) [reply src/L1TokenBridge.sol]: Added the check, as it doesn't hurt to prevent more user errors, even though the standard bridge doesn't block this.

File: src/L1TokenBridge.sol
Comment (oldchili) [inline src/L1TokenBridge.sol]: I see in the standard bridge instead of `isOpen` there is a `paused()` function, which on the l1 is said to check the SuperchainConfig - https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/universal/StandardBridge.sol#L159

Is that of any interest to us? is there anything in the SuperchainConfig that might cause us to want to block something?

Comment (oldchili) [reply src/L1TokenBridge.sol]: That `paused()` actually blocks the finalize operations, so not parallel to our `isOpen`, but still worth understanding.

File: src/L1TokenBridge.sol
Comment (oldchili) [inline src/L1TokenBridge.sol:183]: I see we define different interfaces on the dai bridge, for example - https://github.com/makerdao/optimism-dai-bridge/blob/master/contracts/l1/L1DAITokenBridge.sol#L102

This was a design decision right?
The standard bridge interface was as it is now at the time?

File: src/L1TokenBridge.sol
Comment (telome) [reply src/L1TokenBridge.sol]: I think we probably want it, to get fast [multisig-based](https://docs.optimism.io/chain/security/privileged-roles#guardian) freezing of withdrawals in case of a security incident. Though I think we already get this for free by virtue of using the xchain messenger for withdrawals, see https://github.com/ethereum-optimism/optimism/blob/9001eef4784dc2950d0bdcda29752cb2939bae2b/packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol#L224

Let's keep this comment open for now to remember to add a test that can verify that the flow will indeed be blocked once the OP guardian has paused withdrawals.

File: src/L1TokenBridge.sol
Comment (telome) [reply src/L1TokenBridge.sol:183]: This is the new interface. The dai bridge was using the old interface, pre-bedrock. We might need to also support the old interface as the standard bridge does, but not 100% sure yet, so for now I've only included the methods for the new interface.

File: src/L1TokenBridge.sol
Comment (oldchili) [reply src/L1TokenBridge.sol]: Pretty weird that they check it both in the messenger and in the standard bridge it self, seems like a duplication.

File: src/Escrow.sol
Comment (sunbreak1211) [inline src/Escrow.sol]: nit: i'd suggest to move to the bottom

File: src/L1GovernanceRelay.sol
Comment (sunbreak1211) [inline src/L1GovernanceRelay.sol:37]: I know in the optimism-dai-bridge it wasn't upgradable either (even if not immutable). But are we sure that is totally fine?

File: src/L1TokenBridge.sol
Comment (sunbreak1211) [inline src/L1TokenBridge.sol]: nit: I know this is how the standard bridge does but I'd say to move to the bottom.

File: src/L1TokenBridge.sol
Comment (sunbreak1211) [inline src/L1TokenBridge.sol]: I think I prefer keeping an internal function for both and then two external calling it. Considering that is also the way how the standard bridge is implemented could be nice to do that change.

File: src/L1TokenBridge.sol
Comment (sunbreak1211) [inline src/L1TokenBridge.sol]: Maybe use `encodeCall` to keep better consistency with the GovRelay?

File: src/L2TokenBridge.sol
Comment (sunbreak1211) [inline src/L2TokenBridge.sol]: Same about having an internal and two external.

File: src/L2TokenBridge.sol
Comment (sunbreak1211) [inline src/L2TokenBridge.sol]: Same about moving to the bottom.

File: src/L2TokenBridge.sol
Comment (sunbreak1211) [inline src/L2TokenBridge.sol]: Same about `encodeCall`

File: src/L2TokenBridge.sol
Comment (sunbreak1211) [inline src/L2TokenBridge.sol:1]: Considering both sides of the bridge are almost the same excepting for the mint/burn vs transferFrom escrow, have you considered making an abstract contract for both and just keeping the minimal in the actual L1 and L2 contracts?

Comment (sunbreak1211) [reply src/L2TokenBridge.sol:1]: That could also be a bit more inline to what the standard bridge does (I don't like L2 inheriting from the L1 though).

Comment (sunbreak1211) [reply src/L2TokenBridge.sol:1]: The only annoyance might be the error messages, I guess we would have 2 options:
- just use one standard for both
- have a constant string that we can append as a prefix for the errors.

Comment (oldchili) [reply src/L2TokenBridge.sol:1]: Let's maybe discuss this before making changes.
I'm not totally decided but I kind of like the keeping of separate files as was done in the previous bridge.
Diffing the L1 and L2 files seems pretty straightforward.

BTW the standard bridge actually implements different logic in the same file based on checks if the token is native or not. So even if we use inheritance it will not be exactly similar.

Comment (sunbreak1211) [reply src/L2TokenBridge.sol:1]: Yeah it would still be different. I was picturing an approach like `DssVest`, but yeah I don't feel super strong about it either.

File: src/L2GovernanceRelay.sol
Comment (oldchili) [inline src/L2GovernanceRelay.sol]: Are we sure we want this attempt to retrieve the error string?
The pause-proxy, subdao-proxy and current arbitrum + optimism l2 relays don't have it.
Also we got a comment in one of the audits that it's a bit obsolete, as other revert patterns may be thrown.

File: src/L1GovernanceRelay.sol
Comment (telome) [reply src/L1GovernanceRelay.sol:37]: The same arguments [discussed](https://github.com/makerdao/arbitrum-token-bridge/pull/1#discussion_r1620441951) in the context of the arbitrum-token-bridge still apply here. Generally I think the likelihood of these messenger contracts changing addresses seems low enough that, if it were to occur, it would be fine to run a bridge migration.

File: src/L2GovernanceRelay.sol
Comment (telome) [reply src/L2GovernanceRelay.sol]: I do think it was an oversight not to bubble the revert in the arbitrum/optimism relays as it makes it harder to debug failed L2 spells. But good point re: the decoding, I changed it to use the more generic version suggested by Cantina.

File: src/L1TokenBridge.sol
Comment (telome) [reply src/L1TokenBridge.sol]: Changed

File: src/L1TokenBridge.sol
Comment (telome) [reply src/L1TokenBridge.sol]: Done

File: src/L1TokenBridge.sol
Comment (telome) [reply src/L1TokenBridge.sol]: Done

File: src/L2TokenBridge.sol
Comment (telome) [reply src/L2TokenBridge.sol]: Done

File: src/L2TokenBridge.sol
Comment (telome) [reply src/L2TokenBridge.sol]: Done

File: src/L2TokenBridge.sol
Comment (telome) [reply src/L2TokenBridge.sol]: Done

File: src/L2TokenBridge.sol
Comment (sunbreak1211) [reply src/L2TokenBridge.sol:1]: So I guess it is preferred to keep the duplication in favor to avoid inheritance. Will resolve this excepting we want to discuss something more. cc @telome

File: deploy/mocks/ChainLog.sol
Comment (sunbreak1211) [inline deploy/mocks/ChainLog.sol]:
```suggestion
```

File: test/L2GovernanceRelay.t.sol
Comment (sunbreak1211) [inline test/L2GovernanceRelay.t.sol]: It seems unnecessary.

File: test/L2GovernanceRelay.t.sol
Comment (sunbreak1211) [inline test/L2GovernanceRelay.t.sol:56]: This one too if the prev one is removed.

File: test/L2GovernanceRelay.t.sol
Comment (telome) [reply test/L2GovernanceRelay.t.sol]: This was meant to make it more clear that the first revert occurs due to a wrong msg.sender and the next revert occurs due to a wrong xDomainMessageSender. Have rearranged the test slightly to make it more clear.

Comment (sunbreak1211) [reply test/L2GovernanceRelay.t.sol]: Yeah imagined it was related to that. I was thinking we could have also checked the xDomainMessageSender part first and then the msg.sender one. That could remove the need of the two extra setXDomainMessageSender, but maybe this way is better as being more explicit, so resolving this.

File: test/L1TokenBridge.t.sol
Comment (sunbreak1211) [inline test/L1TokenBridge.t.sol]:
```suggestion
        uint256 eoaBefore = l1Token.balanceOf(address(0xe0a));
```

File: test/L2TokenBridge.t.sol
Comment (sunbreak1211) [inline test/L2TokenBridge.t.sol]:
```suggestion
        uint256 eoaBefore = l2Token.balanceOf(address(0xe0a));
```

File: test/Integration.t.sol
Comment (sunbreak1211) [inline test/Integration.t.sol]:
```suggestion
            l2GovRelay:  l2GovRelay
```
nit

File: test/Integration.t.sol
Comment (sunbreak1211) [inline test/Integration.t.sol]: I'm a bit confused here, the `deployL2` has 3 contract deployments + switching the owner of the bridge, which should make an action as we are indeed moving the owner from the test address to the L2 govRelay. How is it that the nonce needs to increment 3 instead of 4?

File: README.md
Comment (sunbreak1211) [inline README.md]:
```suggestion
- `L2GovernanceRelay.sol` - L2 side of the governance relay.
```

File: test/Integration.t.sol
Comment (sunbreak1211) [reply test/Integration.t.sol]: Already discussed.

File: README.md
Comment (sunbreak1211) [inline README.md]:
```suggestion
To deposit a given amount of a supported token to the L2, Alice calls `bridgeERC20[To]()` on the `L1TokenBridge`. This call locks Alice's tokens into the `Escrow` contract and calls the [L1CrossDomainMessenger](https://github.com/ethereum-optimism/optimism/blob/9001eef4784dc2950d0bdcda29752cb2939bae2b/packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol) which instructs the sequencer to asynchroneously relay a cross-chain message on L2. This will involve a call to `finalizeBridgeERC20()` on `L2TokenBridge`, which mints an equivalent amount of L2 tokens for Alice (or `to`).
```

File: README.md
Comment (sunbreak1211) [inline README.md]:
```suggestion
To withdraw her tokens back to L1, Alice calls `bridgeERC20[To]()` on the `L2TokenBridge`. This call burns Alice's tokens and calls the [L2CrossDomainMessenger](https://github.com/ethereum-optimism/optimism/blob/9001eef4784dc2950d0bdcda29752cb2939bae2b/packages/contracts-bedrock/src/L2/L2CrossDomainMessenger.sol), which will eventually (after the ~7 days security period) allow the permissionless finalization of the withdrawal on L1. This will involve a call to `finalizeBridgeERC20()` on the `L1TokenBridge`, which releases an equivalent amount of L1 tokens from the `Escrow` to Alice (or `to`).
```

Comment (sunbreak1211) [reply README.md]: ^ @telome

File: script/Deploy.s.sol
Comment (sunbreak1211) [inline script/Deploy.s.sol]: This fixed the issue of losing the other contracts in the output for me:
```suggestion
        string memory json = vm.serializeAddress(string(abi.encodePacked(ScriptTools.EXPORT_JSON_KEY, "_", name)), label, addr);
```

I think we should upgrade `dss-test` to the last version in the other 3 repos as well and apply the patch. Or maybe just better to add this to `dss-test` and upgrade it everywhere.

Comment (telome) [reply script/Deploy.s.sol]: Sorry, hadn't seen your suggestion, but I applied the same fix.

Comment (telome) [reply script/Deploy.s.sol]: Added the functions to dss-test https://github.com/makerdao/dss-test/pull/46

Comment (sunbreak1211) [reply script/Deploy.s.sol]: Great! will check the PR.

File: src/L1TokenBridge.sol
Comment (oldchili) [reply src/L1TokenBridge.sol]: We added a check to make sure that withdrawals can be blocked by the guardian, so resolving.

**Review by sunbreak1211 — APPROVED** [review]
Assuming the old interface won't be needed, this looks good to me.

File: README.md
Comment (oldchili) [inline README.md:49]: I see this part is less detailed than https://github.com/makerdao/arbitrum-token-bridge/blob/dev/README.md#upgrade-a-single-token-to-a-new-bridge

Is it because there is some difference of behavior?

Comment (telome) [reply README.md:49]: Yes, the difference is that we are not reading the L2 token from storage during finalization of a deposit for the op bridge, so unregistering the token on the L2 side doesn't risk blocking pending deposits.

File: README.md
Comment (oldchili) [inline README.md]: Consider adding a gas estimate buffer as you suggested here - https://github.com/makerdao/op-farms/pull/1#discussion_r1725100097.

Can use the forge flag I found.

Comment (oldchili) [reply README.md]: TBH it's a bit weird adding it only here as there are L1 => L2 txs being done also in other parts of the scripts. But maybe better than nothing.

**Review by sunbreak1211 — APPROVED** [review]

File: README.md
Comment (telome) [reply README.md]: Done. I'd say it's less of a concern for `Init.s.sol` which is meant to only be used on testnets. But `Deposit.s.sol` could legitimately also be used on mainnet.

**Review by oldchili — APPROVED** [review]
LGTM, also assuming we decide not support the old interface.

**Review by sunbreak1211 — APPROVED** [review]

**Review by oldchili — APPROVED** [review]

**Review by sunbreak1211 — APPROVED** [review]

**Comment by sunbreak1211** [comment]
@oldchili let's give a final approval so @telome can merge it. Should be done with regular merge commit type.

**Review by oldchili — APPROVED** [review]

**Pull request merged**
